src/benchmarklib/benchmark_runner.cpp

```diff
@@ -0,0 +1,449 @@
+#include "planviz/sql_query_plan_visualizer.hpp"
+#include "sql/sql_pipeline_builder.hpp"
+#include "storage/storage_manager.hpp"
+#include "tpch/tpch_db_generator.hpp"
+#include "tpch/tpch_queries.hpp"
+#include "utils/load_table.hpp"
+
+namespace opossum {
+
+BenchmarkRunner::BenchmarkRunner(BenchmarkConfig config, const NamedQueries& queries, const nlohmann::json& context)
+    : _config(std::move(config)), _queries(queries), _context(context) {}
+
+void BenchmarkRunner::run() {
+  _config.out << "\n- Starting Benchmark.." << std::endl;
+
+  // Run the queries in the selected mode
+  switch (_config.benchmark_mode) {
+    case BenchmarkMode::IndividualQueries: {
+      _benchmark_individual_queries();
+      break;
+    }
+    case BenchmarkMode::PermutedQuerySets: {
+      _benchmark_permuted_query_sets();
+      break;
+    }
+  }
+
+  // Create report
+  if (_config.output_file_path) {
+    std::ofstream output_file(*_config.output_file_path);
+    _create_report(output_file);
+  } else {
+    _create_report(std::cout);
+  }
+
+  // Visualize query plans
+  if (_config.enable_visualization) {
+    for (const auto& name_and_plans : _query_plans) {
+      const auto& name = name_and_plans.first;
+      const auto& lqps = name_and_plans.second.lqps;
+      const auto& pqps = name_and_plans.second.pqps;
+
+      GraphvizConfig graphviz_config;
+      graphviz_config.format = "svg";
+
+      for (auto lqp_idx = size_t{0}; lqp_idx < lqps.size(); ++lqp_idx) {
+        const auto file_prefix = name + "-LQP-" + std::to_string(lqp_idx);
+        LQPVisualizer{graphviz_config, {}, {}, {}}.visualize({lqps[lqp_idx]}, file_prefix + ".dot",
+                                                             file_prefix + ".svg");
+      }
+      for (auto pqp_idx = size_t{0}; pqp_idx < pqps.size(); ++pqp_idx) {
+        const auto file_prefix = name + "-PQP-" + std::to_string(pqp_idx);
+        SQLQueryPlanVisualizer{graphviz_config, {}, {}, {}}.visualize(*pqps[pqp_idx], file_prefix + ".dot",
+                                                                      file_prefix + ".svg");
+      }
+    }
+  }
+}
+
+void BenchmarkRunner::_benchmark_permuted_query_sets() {
+  // Init results
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    _query_results_by_query_name.emplace(name, QueryBenchmarkResult{});
+  }
+
+  auto mutable_named_queries = _queries;
+
+  // For shuffling the query order
+  std::random_device random_device;
+  std::mt19937 random_generator(random_device());
+
+  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
+  while (state.keep_running()) {
+    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+
+    for (const auto& named_query : mutable_named_queries) {
+      const auto query_benchmark_begin = std::chrono::steady_clock::now();
+
+      // Execute the query, we don't care about the results
+      _execute_query(named_query);
```

Comment:
`benchmark::DoNotOptimize`?

Reply:
is that a feature request or something that i missed? again, this is just copied from the TPCH benchmark.

Reply:
I didn't check what was copied and what not. That would be the google-benchmarky way to express that comment